### PR TITLE
Micro rtps send poll

### PR DESCRIPTION
**Pull request by PepMS**
Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary. You can then open draft pull requests to get early feedback.

**Describe problem solved by this pull request**
micrRTPS client has subscribers to the topics we want to send to ROS2. However, the client cannot keep up the rates for high rate topics (e.g. `vehicle_angular_velocity` or `vehicle_attitude`). One solution to this problem was to remove the `px4_usleep` at [this line](https://github.com/PX4/PX4-Autopilot/blob/29616ec387b97be3a384e1b35d5306de1b6c6c2a/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L182) but by doing this the CPU usage skyrockets.

This PR solves the problem of sending data of high rate topics through RTPS without a critical increase in CPU usage.

**Describe your solution**
As briefly discussed with @TSC21 and @bkueng in the Slack channel, the solution is to poll from high rate topics by specifying this in the yaml file.

Additionally, I've added a field `poll_interval` to specify the maximum polling period of a topic.

**Describe possible alternatives**
Another possible option would be to have a px4_usleep() with a smaller granularity than 1ms, so that we could reduce the sleep period.

**Test data / coverage**
Using SITL and specifying several topics to send (some with the poll option and some without). The results are the expected.

Also, Pixhwak 4 Mini. I've managed to send `vehicle_angular_velocity`, `vehicle_attitude`, `vehicle_position` at 250Hz, 200Hz, and 100Hz respectively. The CPU usage went from 68% to 77% approx.

**Additional context**
I am not sure what would happen if we try to poll from low rates topics. I guess that the micrRTPS client sending thread would be sleeping until it receives some data on this topic. Meaning that all other messages in the other sent topics would be delayed by this one.
As a rule of thumb, I would advise that if the poll option is enabled, the topic with the highest rate should have the polling activated too.

### CODE REVIEW

**Comment by llanesc** [comment]
Would poll be a redundant key here? Could you just have the poll_interval key and if it exists then poll = true.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:226]: This should be checked in any case, not only on poll timeout.
Also can you avoid the code duplication and move this to a separate function (or move it outside of `@[ if send_topics_poll]@`)?

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:210]: There should be a sleep here, to avoid busy-looping

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: You can remove this here.
For the non-polled topics you can add a check if `tx_interval` time has passed since the last send, to avoid increasing their rates if there's a fast polled topic.

File: msg/tools/px_generate_uorb_topic_helper.py
Comment (bkueng) [inline msg/tools/px_generate_uorb_topic_helper.py:441]:
```suggestion
            return dict.get('poll_interval', False)
```

File: msg/tools/urtps_bridge_topics.yaml
Comment (bkueng) [inline msg/tools/urtps_bridge_topics.yaml:31]: Was this just for testing? Can you restore the file?

File: msg/tools/uorb_rtps_classifier.py
Comment (bkueng) [inline msg/tools/uorb_rtps_classifier.py:69]:
```suggestion
                poll = topic.get('poll', False)
```
And the same below

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: I'm not sure I understand what you mean, sorry.  To avoid increasing the non-polled topics rates you mean? If this is what you mean, I wouldn't be increasing their rates since I'm calling update and only send this topic in case the update is true. Is that correct?

Comment (bkueng) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: > Do you mean to avoid increasing the non-polled topics rates?

Yes. The current behavior is: you poll on a topic and then run a full update loop. This means if your polled topic updates at a fast rate (and you don't restrict it via poll_interval), all others will be as well, without considering `tx_interval`. Does that make sense?

Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: Ok, so I may have a misunderstanding of the `poll` function. 

The way I see it for the **polled topics**:
1. Here, we poll all poll-topics inside the fds https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L207 
2. Then, we only send through RTPS the topics that have been published during the `poll` sleep. We distinguish among the the "updated" topics with this condition: https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L213 I am not sure whether with this condition we can check that. For the tests I did, it seemed to me that we can.

If this is true, then we would not need any sleep, right? Since we would only be publishing the messages that have been updated during the `poll` call. Thus every polled topic will be sent, at maximum, at the same rate it is being published in the uORB topic.

For the **non-polled topics**:
1. Before calling to `poll` we update all non-polled topics here: https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L169
2. If there's new data in the topic this function will return true and we will send it via RTPS. Otherwise, we continue.
3. If this is true, then we cannot send this topic at higher rates, since this function will return false if there are no new data.

Sorry for this long message. I wanted to be sure that I correctly understood the functionality behind each method.

Comment (bkueng) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: > If this is true, then we would not need any sleep, right?

Yes, correct.

> If this is true, then we cannot send this topic at higher rates, since this function will return false if there are no new data.

Yes correct as well.
I was refering to a case like this:
- `sensor_accel` polled, updated at 250Hz
- `vehicle_attitude` not polled. also updated at 250Hz. But assume we don't want to send it at 250Hz, then the above logic does not allow for it. Therefore I suggest to include a check for `tx_interval`, so that if you set it to very small (e.g. 1ms), you can get attitude at 250Hz as well, but you can increase the interval to reduce the sending rate of the attitude only, and still get sensor_accel at full rate.

Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: Ah ok! now I see the case that you meant. However, I'm not sure how to implement it. It seems to me that if we don't want to delay the whole loop we have to check `tx_interval` before this line:
https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L169

Would you mind giving me some hints on how to implement this?

Comment (bkueng) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:238]: Exactly, a very simple way (with some inaccuracies) is be to add a check before https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L165
```
hrt_abstime last_update = 0;
while (!_should_exit_task) {
  hrt_abstime now = hrt_absolute_time();
  if (now - last_update > tx_interval) {
    last_update = now;
    ...
```

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em]: Done

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:210]: Done. To not hardcode a value in the sleep period I've used `tx_interval`.

File: msg/tools/px_generate_uorb_topic_helper.py
Comment (PepMS) [reply msg/tools/px_generate_uorb_topic_helper.py]: I've removed these methods from the helper and left this file as it was. These methods were used with the legacy RTPS bridge code.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em
Comment (PepMS) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em]: Ah ok, the simple way indeed. Thanks for this. Done.

File: msg/tools/urtps_bridge_topics.yaml
Comment (PepMS) [reply msg/tools/urtps_bridge_topics.yaml]: Yes, correct. The file has been restored.

File: msg/tools/uorb_rtps_classifier.py
Comment (PepMS) [reply msg/tools/uorb_rtps_classifier.py]: Done

**Comment by PepMS** [comment]
> Would poll be a redundant key here? Could you just have the poll_interval key and if it exists then poll = true.

Indeed, it is redundant. However, as the `poll_intervall` is an optional key, I've preferred to leave the `poll` key for the sake of consistency (and because makes things simpler when separating polled from non-polled topics).

**Review by bkueng — DISMISSED** [review]
Looks good. Waiting for CI to pass.

**Comment by TSC21** [comment]
> Looks good. Waiting for CI to pass.

@bkueng before merging I would like to have a review as well still during this week. Thanks

**Comment by TSC21** [comment]
@PepMS can you rebase meanwhile? Thanks

**Comment by PepMS** [comment]
Rebase done, @TSC21

**Comment by bkueng** [comment]
>  @bkueng before merging I would like to have a review as well still during this week. Thanks

Sure, I'll let you do the final merge/review.

**Comment by PepMS** [comment]
@TSC21  , any updates?

**Comment by dagar** [comment]
Closing as obsolete (we can comfortably get high rate topics with the new microdds_client and updates https://github.com/PX4/PX4-Autopilot/pull/20227).

**Pull request closed without merging**
